Libraries/LibMarkdown/Table.cpp

```diff
@@ -0,0 +1,221 @@
+        builder.append("<th>");
+        builder.append(column.header.render_to_html());
+        builder.append("</th>");
+    }
+    builder.append("</tr>");
+    builder.append("</thead>");
+    builder.append("<tbody>");
+    for (size_t i = 0; i < m_row_count; ++i) {
+        builder.append("<tr>");
+        for (auto& column : m_columns) {
+            if (column.rows.size() > i) {
+                builder.append("<td>");
+                builder.append(column.rows[i].render_to_html());
+                builder.append("</td>");
+            }
+        }
+        builder.append("</tr>");
+    }
+    builder.append("</tbody>");
+    builder.append("</table>");
+
+    return builder.to_string();
+}
+
+OwnPtr<Table> Table::parse(Vector<StringView>::ConstIterator& lines)
+{
+    auto peek_it = lines;
+    auto first_line = *peek_it;
+    if (!first_line.starts_with('|'))
+        return nullptr;
+
+    ++peek_it;
+
+    if (peek_it.is_end())
+        return nullptr;
+
+    auto header_segments = first_line.split_view('|', true);
+    auto header_delimiters = peek_it->split_view('|', true);
+
+    header_segments.take_first();
+    header_delimiters.take_first();
+    if (header_segments.last().is_empty())
+        header_segments.take_last();
+    if (header_delimiters.last().is_empty())
+        header_delimiters.take_last();
+
+    ++peek_it;
+
+    if (header_delimiters.size() != header_segments.size())
+        return nullptr;
+
+    if (header_delimiters.is_empty())
+        return nullptr;
+
+    size_t total_width = 0;
+
+    auto table = make<Table>();
+    table->m_columns.resize(header_delimiters.size());
+
+    for (size_t i = 0; i < header_segments.size(); ++i) {
+        auto text_option = Text::parse(header_segments[i]);
+        ASSERT(text_option.has_value()); // Text::parse() cannot fail yet returns an Optional. why?
+        auto text = text_option.release_value();
+        auto& column = table->m_columns[i];
+
+        column.header = move(text);
+
+        auto& delimiter = header_delimiters[i];
+
+        while (delimiter.starts_with(' '))
+            delimiter = delimiter.substring_view(1, delimiter.length() - 1);
+        while (delimiter.ends_with(' '))
+            delimiter = delimiter.substring_view(0, delimiter.length() - 1);
+
+        auto left = delimiter.starts_with(':');
+        auto right = delimiter != ":" && delimiter.ends_with(':');
+
+        if (left)
+            delimiter = delimiter.substring_view(1, delimiter.length() - 1);
+        if (right)
+            delimiter = delimiter.substring_view(0, delimiter.length() - 1);
+
+        if (left && right)
+            column.alignment = Alignment::Center;
+        else if (right)
+            column.alignment = Alignment::Right;
+
+        size_t relative_width = delimiter.length();
+        for (auto ch : delimiter) {
+            if (ch != '-') {
+                dbg() << "Invalid character '" << ch << "' in table heading delimiter";
```

Comment:
I *think* the idiomatic *Serenity way* to quote weird characters is with a pair of \_'s instead of single quotes, but you might want to ask @awesomekling about that.

Reply:
I've seen that too, sure, will use pair of _'s.